keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,548 @@
+1. Feature has been stable in lifecyle `PreRelease: Deprecated, {Default: defaultVal, LockToDefault: true` for at least 1 minor release.
+1. Feature can be removed in the next release.
+
+The figure below shows an example of how a GA feature can be removed.
+![ga feature removal](./etcd_ga_feature_removal.png "ga feature removal")
+
+To ensure this works in mixed version scenarios, we require:
+1. The `LockToDefault` state of a feature cannot be changed in patch version.
+1. The `LockToDefault` state of a feature can only be set to be `true` if the feature has been stable for at least 1 minor version, which means no change of its default value in any patch version of the previous minor version.
+1. The server would need to use the default value of the feature if `LockToDefault=true` even when the cluster feature gate is not set.
+
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+New feature gate unit tests will be added.
+
+##### Integration tests
+
+New feature gate integration tests will be added.
+
+##### e2e tests
+
+We will add one cluster level experimental features to the feature gate (without removing the original experimental flag), and add e2e tests to make sure the feature gate functions equivalently to their `--experimental-xxx` flags.
+
+We will also add downgrade/upgrade e2e tests to make sure the feature gate does not break the cluster in mixed version cluster.
+
+### Graduation Criteria
+
+#### Milestone 1
+
+* cluster level feature gate implemented.
+  * new raft changes added.
+  * cluster feature gate added to the server code, and used by a cluster level experimental feature.
+  * feature metrics added.
+  * e2e tests added for the feature gate equivalent of the selected experimental feature.
+  * robustness tests added for current version.
+
+#### Milestone 2
+* cluster level feature gate client API implemented.
+  * grpc endpoints added.
+  * `clientv3` function added.
+  * `etcdctl` command added.
+
+#### Milestone 3
+
+* cluster level feature gate fully tested.
+  * backport necessary miminal changes to 3.5.
+  * feature gates clean up added to `schema.UnsafeMigrate`.
+  * e2e tests for upgrade & downgrade & mixed version scenarios completed.
+  * robustness tests with upgrade & downgrade & mixed version scenarios completed. 
+  * documentation.
+
+### Upgrade / Downgrade Strategy
+
+For regular downgrade process, `schema.UnsafeMigrate` should clean up feature gates that are not available in the target version.
+
+The feature gate feature would available in 3.6+. 
+In 3.5, the necessary proto changes and no-op raft changes would be backported, but feature gate would not be available. 
```

Comment:
Could you just implement it via on a client side? If server returns notImplemented(404) for the method, we can assume it doesn't implement any features.